### PR TITLE
add CSRF validation by custom HTTP header

### PR DESCRIPTION
**Pull request by olegbaturin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #20279

### CODE REVIEW

**Comment by samdark** [comment]
This is backport from Yii3. Good one.

**Review by samdark — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

File: framework/web/Request.php [hunk 1]
```diff
@@ -1860,8 +1895,14 @@ protected function createCsrfCookie($token)
     public function validateCsrfToken($clientSuppliedToken = null)
     {
         $method = $this->getMethod();
-        // only validate CSRF token on non-"safe" methods https://tools.ietf.org/html/rfc2616#section-9.1.1
-        if (!$this->enableCsrfValidation || in_array($method, ['GET', 'HEAD', 'OPTIONS'], true)) {
+
+        if ($this->validateCsrfHeaderOnly) {
+            return in_array($method, $this->csrfHeaderUnafeMethods, true)
```
Comment (bizley) [inline framework/web/Request.php:1900]:
```suggestion
            return in_array($method, $this->csrfHeaderUnsafeMethods, true)
```

**Review by bizley — CHANGES_REQUESTED** [review]

**Comment by mtangoo** [comment]
I think this PR is good with exception of typos as pointed out by @bizley 
@olegbaturin would you please address them?

**Review by bizley — APPROVED** [review]

**Comment by mtangoo** [comment]
@olegbaturin thanks for your time and efforts

**Pull request merged**